### PR TITLE
FIX [`PEFT` / `Trainer` ] Handle better peft + quantized compiled models

### LINKED ISSUE #29033: Trainer doesn't handle torch.compiled QLoRA models correctly
### System Info

- `transformers` version: 4.38.0.dev0
- Platform: Linux-6.1.58+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: 0.28.0.dev0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- Tensorflow version (GPU?): 2.15.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.8.1 (cpu)
- Jax version: 0.4.23
- JaxLib version: 0.4.23
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
!pip install bitsandbytes
!pip install git+https://github.com/huggingface/accelerate
!pip install git+https://github.com/huggingface/datasets
!pip install git+https://github.com/huggingface/peft
!pip install git+https://github.com/huggingface/transformers
!pip install git+https://github.com/huggingface/trl

import torch

import accelerate
import datasets
import peft
import transformers
import trl

import bitsandbytes

train_dataset = datasets.load_dataset('imdb', split='train')

bnb_config = transformers.BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_quant_type="nf4",
)

lora_config = peft.LoraConfig(
    r=8,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM"
)

tokenizer = transformers.AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")
model = transformers.AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-v0.1", quantization_config=bnb_config)
model = peft.prepare_model_for_kbit_training(model)
model = peft.get_peft_model(model, lora_config)

trainer = trl.SFTTrainer(
    # model=model,  # Does not raise a ValueError
    model=torch.compile(model),  # Raises a ValueError
    train_dataset=train_dataset,
    dataset_text_field='text',
    max_seq_length=512,
)
```

### Expected behavior

Expected Behaviour: 

Best Case: Calling torch.compile has no effect on whether an exception is raised.
Worst Case: Raising an exception that reflects that torch.compile isn't supported.

Current behaviour: 
```

[/usr/local/lib/python3.10/dist-packages/transformers/trainer.py](https://localhost:8080/#) in __init__(self, model, args, data_collator, train_dataset, eval_dataset, tokenizer, model_init, compute_metrics, callbacks, optimizers, preprocess_logits_for_metrics)
    431         # At this stage the model is already loaded
    432         if _is_quantized_and_base_model and not _is_peft_model(model):
--> 433             raise ValueError(
    434                 "You cannot perform fine-tuning on purely quantized models. Please attach trainable adapters on top of"
    435                 " the quantized model to correctly perform fine-tuning. Please see: https://huggingface.co/docs/transformers/peft"

ValueError: You cannot perform fine-tuning on purely quantized models. Please attach trainable adapters on top of the quantized model to correctly perform fine-tuning. Please see: https://huggingface.co/docs/transformers/peft for more details

```

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes: https://github.com/huggingface/transformers/issues/29033

Even though quantized models + compile + peft is not really stable (might not work OTB for all users), the current way we deal with peft compiled models leads to errors that are hard to interpret to users such as the one described in https://github.com/huggingface/transformers/issues/29033 

I will run some tests on my see to check if torch.compile + qlora is supported

cc @amyeroberts

### CODE REVIEW

**Comment by amyeroberts** [comment]
LGTM! Let me know when it's out of draft and you want a final review

**Comment by amyeroberts** [comment]
It would be good to add a test for compiled models alongside this change

**Comment by younesbelkada** [comment]
Thanks @amyeroberts ! This is ready for review ! 
Note currently QLoRA + torch.compile is not really supported as we need to make some changes to make 4bit layers from bnb compatible with bitsandbytes. Therefore I added a test that simply checks if the trainer is correctly initialized so that we don't get the unappropriate error from #29033 . Once compile + QLoRA will be supported, this will be seemless as users will simply have to update `bitsandbytes` to make it work

File: tests/trainer/test_trainer.py
Comment (amyeroberts) [inline tests/trainer/test_trainer.py:913]: If this is the case - we should raise an explicit error if using bnb + compile with trainer.

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing and adding a test! 

If it's still not yet possible to use a compiled model with bnb, then we should raise an error saying this to the user instead of letting things fail when they call `.train`

**Comment by younesbelkada** [comment]
Thanks @amyeroberts for the review! I just adapted the PR according to your suggestion, one concern I have with this new approach is that if bnb supports `torch.compile` in the next months we'll need to do some version check on trainer's init and make the code slightly bloated - though I am happy with this approach as well as I don't think this is that bad. What do you think? 🙏

**Comment by amyeroberts** [comment]
@younesbelkada How soon and with what confidence do you think bnb will support this? 

 I agree we want to avoid bloat as much as possible, but I think it's better overall to be explict and have well defined behaviour for our users. The easiest way to avoid many version checks is to only support the latest version of bnb.

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

**Comment by younesbelkada** [comment]
OK sounds good! I agree with that approach!
Currently we don't have any ETA :/ it might take a bit long, and it will also depend if the community will ask for it

**Comment by pharringtonp19** [comment]
@younesbelkada I would certainly be interested if bnb could support torch.compile!

**Pull request merged**
